File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseUserRepository.java
package com.yahyeet.boardbook.model.firebase.repository;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;
import com.yahyeet.boardbook.model.entity.User;
import com.yahyeet.boardbook.model.repository.IUserRepository;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.stream.Collectors;

/**
 * Firebase Firestore implementation of a user repository
 */
public class FirebaseUserRepository extends AbstractFirebaseRepository<User> implements IUserRepository {

	public FirebaseUserRepository(FirebaseFirestore firestore) {
		super(firestore);
	}

	@Override
	public AbstractFirebaseEntity<User> fromModelEntityToFirebaseEntity(User entity) {
		return FirebaseUser.fromModelType(entity);
	}

	@Override
	public AbstractFirebaseEntity<User> fromDocumentToFirebaseEntity(DocumentSnapshot document) {
		return FirebaseUser.fromDocument(document);
	}

	@Override
	public String getCollectionName() {
		return "users";
	}

	@Override
	public CompletableFuture<List<User>> findFriendsByUserId(String id) {
		CompletableFuture<List<User>> friendsPartOne = CompletableFuture.supplyAsync(() -> {
			Task<QuerySnapshot> task =
				getFirestore()
					.collection(getFriendsCollectionName())
					.whereEqualTo("a", id)
					.get();

			try {
				QuerySnapshot querySnapshot = Tasks.await(task);

				return querySnapshot
					.getDocuments()
					.stream()
					.map(document -> document.getString("b"))
					.collect(Collectors.toList());
			} catch (Exception e) {
				throw new CompletionException(e);
			}
		}).thenCompose(friendIds -> {
			List<CompletableFuture<User>> friendsFuture = friendIds
				.stream()
				.map(this::find)
				.collect(Collectors.toList());

			CompletableFuture<Void> allOfFriendsFuture = CompletableFuture.allOf(friendsFuture.toArray(new CompletableFuture[0]));

			return allOfFriendsFuture.thenApply(future -> friendsFuture.stream().map(CompletableFuture::join).collect(Collectors.toList()));
		});

		CompletableFuture<List<User>> friendsPartTwo = CompletableFuture.supplyAsync(() -> {
			Task<QuerySnapshot> task =
				getFirestore()
					.collection(getFriendsCollectionName())
					.whereEqualTo("b", id)
					.get();

			try {
				QuerySnapshot querySnapshot = Tasks.await(task);

				return querySnapshot
					.getDocuments()
					.stream()
					.map(document -> document.getString("a"))
					.collect(Collectors.toList());
			} catch (Exception e) {
				throw new CompletionException(e);
			}
		}).thenCompose(friendIds -> {
			List<CompletableFuture<User>> friendsFuture = friendIds
				.stream()
				.map(friendId -> find(friendId))
				.collect(Collectors.toList());

			CompletableFuture<Void> allOfFriendsFuture = CompletableFuture.allOf(friendsFuture.toArray(new CompletableFuture[0]));

			return allOfFriendsFuture.thenApply(future -> friendsFuture.stream().map(CompletableFuture::join).collect(Collectors.toList()));
		});

		return friendsPartOne
			.thenCombine(friendsPartTwo, (partOne, partTwo) -> {
				partOne.addAll(partTwo);
				return partOne;
			});
	}

	@Override
	public CompletableFuture<Void> afterSave(User entity, AbstractFirebaseEntity<User> savedEntity) {
		return findFriendsByUserId(savedEntity.getId()).thenApply(remoteFriends ->
			entity
				.getFriends()
				.stream()
				.filter(localFriend ->
					remoteFriends
						.stream()
						.noneMatch(remoteFriend -> remoteFriend.equals(localFriend))
				)
				.collect(Collectors.toList()
				))
			.thenCompose(addedFriends -> {
				List<CompletableFuture<Void>> addedFriendsFuture = addedFriends
					.stream()
					.map(addedFriend -> addFriend(entity.getId(), addedFriend.getId()))
					.collect(Collectors.toList());

				return CompletableFuture.allOf(addedFriendsFuture.toArray(new CompletableFuture[0]));
			});
	}

	/**
	 * Connects two user in a pivot collection in Firebase Firestore
	 *
	 * @param leftId First user
	 * @param rightId Second user
	 * @return A completable future that when resolves denotes that the operation is finished, if the
	 * operation is unsuccessful it throws an error
	 */
	private CompletableFuture<Void> addFriend(String leftId, String rightId) {
		return CompletableFuture.supplyAsync(() -> {
			Map<String, Object> data = new HashMap<>();
			data.put("a", leftId);
			data.put("b", rightId);
			Task<DocumentReference> task = getFirestore().collection(getFriendsCollectionName()).add(data);

			try {
				Tasks.await(task);

				return null;
			} catch (Exception e) {
				throw new CompletionException(e);
			}
		});
	}

	private String getFriendsCollectionName() {
		return getCollectionNamePrefix() + "_friends_user_to_user";
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseUserRepository.kt
package com.yahyeet.boardbook.model.firebase.repository

import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import com.yahyeet.boardbook.model.entity.User
import com.yahyeet.boardbook.model.repository.IUserRepository

import java.util.HashMap
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.stream.Collectors

class FirebaseUserRepository(firestore: FirebaseFirestore) : AbstractFirebaseRepository<User>(firestore), IUserRepository {

    override val collectionName: String
        get() = "users"

    private val friendsCollectionName: String
        get() = collectionNamePrefix + "_friends_user_to_user"

    override fun fromModelEntityToFirebaseEntity(entity: User): AbstractFirebaseEntity<User> {
        return FirebaseUser.fromModelType(entity)
    }

    override fun fromDocumentToFirebaseEntity(document: DocumentSnapshot): AbstractFirebaseEntity<User> {
        return FirebaseUser.fromDocument(document)
    }

    override fun findFriendsByUserId(id: String?): CompletableFuture<List<User>> {
        val friendsPartOne = CompletableFuture.supplyAsync<List<String>> {
            val task = firestore
                    .collection(friendsCollectionName)
                    .whereEqualTo("a", id)
                    .get()

            try {
                val querySnapshot = Tasks.await(task)

                return@CompletableFuture.supplyAsync querySnapshot
                        .getDocuments()
                        .stream()
                        .map({ document -> document.getString("b") })
                        .collect(Collectors.toList<String>())
            } catch (e: Exception) {
                throw CompletionException(e)
            }
        }.thenCompose { friendIds ->
            val friendsFuture = friendIds
                    .stream()
                    .map<CompletableFuture<User>>(Function<String, CompletableFuture<User>> { this.find(it) })
                    .collect<List<CompletableFuture<User>>, Any>(Collectors.toList())

            val allOfFriendsFuture = CompletableFuture.allOf(*friendsFuture.toTypedArray<CompletableFuture>())

            allOfFriendsFuture.thenApply { future -> friendsFuture.stream().map<User>(Function<CompletableFuture<User>, User> { it.join() }).collect<List<User>, Any>(Collectors.toList()) }
        }

        val friendsPartTwo = CompletableFuture.supplyAsync<List<String>> {
            val task = firestore
                    .collection(friendsCollectionName)
                    .whereEqualTo("b", id)
                    .get()

            try {
                val querySnapshot = Tasks.await(task)

                return@CompletableFuture.supplyAsync querySnapshot
                        .getDocuments()
                        .stream()
                        .map({ document -> document.getString("a") })
                        .collect(Collectors.toList<String>())
            } catch (e: Exception) {
                throw CompletionException(e)
            }
        }.thenCompose { friendIds ->
            val friendsFuture = friendIds
                    .stream()
                    .map { friendId -> find(friendId) }
                    .collect<List<CompletableFuture<User>>, Any>(Collectors.toList())

            val allOfFriendsFuture = CompletableFuture.allOf(*friendsFuture.toTypedArray<CompletableFuture>())

            allOfFriendsFuture.thenApply { future -> friendsFuture.stream().map<User>(Function<CompletableFuture<User>, User> { it.join() }).collect<List<User>, Any>(Collectors.toList()) }
        }

        return friendsPartOne
                .thenCombine(friendsPartTwo) { partOne, partTwo ->
                    partOne.addAll(partTwo)
                    partOne
                }
    }

    override fun afterSave(entity: User, savedEntity: AbstractFirebaseEntity<User>): CompletableFuture<Void> {
        return findFriendsByUserId(savedEntity.id).thenApply { remoteFriends ->
            entity
                    .friends
                    .stream()
                    .filter { localFriend ->
                        remoteFriends
                                .stream()
                                .noneMatch { remoteFriend -> remoteFriend == localFriend }
                    }
                    .collect<List<User>, Any>(Collectors.toList()
                    )
        }
                .thenCompose { addedFriends ->
                    val addedFriendsFuture = addedFriends
                            .stream()
                            .map { addedFriend -> addFriend(entity.id, addedFriend.id) }
                            .collect<List<CompletableFuture<Void>>, Any>(Collectors.toList())

                    CompletableFuture.allOf(*addedFriendsFuture.toTypedArray<CompletableFuture>())
                }
    }

    private fun addFriend(leftId: String?, rightId: String?): CompletableFuture<Void> {
        return CompletableFuture.supplyAsync {
            val data = HashMap<String, Any>()
            data["a"] = leftId!!
            data["b"] = rightId!!
            val task = firestore.collection(friendsCollectionName).add(data)

            try {
                Tasks.await(task)

                return@CompletableFuture.supplyAsync null
            } catch (e: Exception) {
                throw CompletionException(e)
            }
        }
    }
}
